File: frontend/src/types/weather.ts
export interface City {
  id: number;
  department_id: number;
  name: string;
  province?: string;
  latitude: number;
  longitude: number;
  altitude: number;
  is_featured: boolean;
  is_capital: boolean;
  department_name?: string;
  region_natural?: 'Costa' | 'Sierra' | 'Selva' | string;
}

export interface Department {
  id: number;
  name: string;
  code: string;
  capital: string;
  latitude: number;
  longitude: number;
  region_natural: string;
  description?: string;
  cities: City[];
}

export interface CurrentWeather {
  temperature: number;
  apparent_temperature: number;
  relative_humidity: number;
  wind_speed: number;
  wind_direction: number;
  wind_gusts?: number | null;
  surface_pressure: number;
  precipitation: number;
  precipitation_probability?: number | null;
  cloud_cover: number;
  uv_index: number;
  uv_category: string;
  weather_code: number;
  weather_description: string;
  weather_icon: string;
  is_day: boolean;
  temp_max: number;
  temp_min: number;
  sunrise?: string;
  sunset?: string;
  updated_at: string;
  city_id?: number | null;
  city_name: string;
  department_name?: string | null;
  region_natural?: string | null;
  altitude?: number | null;
  latitude: number;
  longitude: number;
}

export interface HourlyForecastItem {
  time: string;
  hour_label: string;
  temperature: number;
  apparent_temperature: number;
  relative_humidity: number;
  precipitation_probability: number;
  precipitation: number;
  weather_code: number;
  weather_description: string;
  weather_icon: string;
  wind_speed: number;
  uv_index: number;
  is_day: boolean;
}

export interface DailyForecastItem {
  date: string;
  day_name: string;
  day_short: string;
  temp_max: number;
  temp_min: number;
  weather_code: number;
  weather_description: string;
  weather_icon: string;
  precipitation_sum: number;
  precipitation_probability_max: number;
  uv_index_max: number;
  wind_speed_max: number;
  sunrise: string;
  sunset: string;
}

export interface FullForecastResponse {
  current: CurrentWeather;
  hourly: HourlyForecastItem[];
  daily: DailyForecastItem[];
}

export interface DepartmentWeatherSummary {
  department_id: number;
  department_name: string;
  capital: string;
  latitude: number;
  longitude: number;
  region_natural: string;
  temperature: number;
  weather_description: string;
  weather_icon: string;
  relative_humidity: number;
  precipitation: number;
  uv_index: number;
  wind_speed: number;
}

export interface WeatherAlert {
  id: string;
  city_id?: number | null;
  city_name: string;
  department_name?: string | null;
  region_natural?: string | null;
  title: string;
  severity: 'info' | 'caution' | 'warning' | 'danger';
  severity_label: string;
  category: string;
  icon: string;
  description: string;
  recommendation: string;
  trigger_value: string;
  threshold: string;
  timestamp: string;
}

export interface AlertsResponse {
  total_alerts: number;
  danger_count: number;
  warning_count: number;
  caution_count: number;
  info_count: number;
  alerts: WeatherAlert[];
}

export interface CityComparisonItem {
  city_id: number;
  city_name: string;
  department_name: string;
  region_natural: string;
  altitude: number;
  temperature: number;
  apparent_temperature: number;
  relative_humidity: number;
  precipitation: number;
  wind_speed: number;
  uv_index: number;
  temp_max: number;
  temp_min: number;
  weather_description: string;
  weather_icon: string;
  surface_pressure: number;
}

export interface CompareResponse {
  count: number;
  cities: CityComparisonItem[];
}

export interface HistoryDataPoint {
  date: string;
  temp_max?: number;
  temp_min?: number;
  temp_mean?: number;
  precipitation_sum?: number;
  wind_speed_max?: number;
  relative_humidity_mean?: number;
  weather_code?: number;
}

export interface HistoryStats {
  average: number;
  maximum: number;
  minimum: number;
  trend: 'ascendente' | 'descendente' | 'estable' | string;
  total_precipitation?: number;
  days_analyzed: number;
}

export interface HistoryResponse {
  city_id: number;
  city_name: string;
  department_name: string;
  variable: string;
  start_date: string;
  end_date: string;
  stats: HistoryStats;
  data: HistoryDataPoint[];
}

export interface NationalRankings {
  hottest: Array<Record<string, any>>;
  coldest: Array<Record<string, any>>;
  highest_uv: Array<Record<string, any>>;
  rainiest: Array<Record<string, any>>;
  windiest: Array<Record<string, any>>;
  total_cities_evaluated: number;
}

File: api/app/schemas/weather.py
from pydantic import BaseModel
from typing import Optional, List

class WeatherCondition(BaseModel):
    code: int
    description: str
    icon: str # icon code or name e.g. "sun", "cloud-rain", "cloud-lightning"

class CurrentWeather(BaseModel):
    temperature: float
    apparent_temperature: float
    relative_humidity: int
    wind_speed: float
    wind_direction: int
    wind_gusts: Optional[float] = None
    surface_pressure: float
    precipitation: float
    precipitation_probability: Optional[int] = None
    cloud_cover: int
    uv_index: float
    uv_category: str # "Bajo", "Moderado", "Alto", "Muy Alto", "Extremo"
    weather_code: int
    weather_description: str
    weather_icon: str
    is_day: bool
    temp_max: float
    temp_min: float
    sunrise: Optional[str] = None
    sunset: Optional[str] = None
    updated_at: str
    city_id: Optional[int] = None
    city_name: str
    department_name: Optional[str] = None
    region_natural: Optional[str] = None
    altitude: Optional[int] = None
    latitude: float
    longitude: float

class HourlyForecastItem(BaseModel):
    time: str # "2026-08-24T14:00"
    hour_label: str # "14:00"
    temperature: float
    apparent_temperature: float
    relative_humidity: int
    precipitation_probability: int
    precipitation: float
    weather_code: int
    weather_description: str
    weather_icon: str
    wind_speed: float
    uv_index: float
    is_day: bool

class DailyForecastItem(BaseModel):
    date: str # "2026-08-24"
    day_name: str # "Lunes", "Martes"
    day_short: str # "Lun", "Mar"
    temp_max: float
    temp_min: float
    weather_code: int
    weather_description: str
    weather_icon: str
    precipitation_sum: float
    precipitation_probability_max: int
    uv_index_max: float
    wind_speed_max: float
    sunrise: str
    sunset: str

class FullForecastResponse(BaseModel):
    current: CurrentWeather
    hourly: List[HourlyForecastItem]
    daily: List[DailyForecastItem]

class DepartmentWeatherSummary(BaseModel):
    department_id: int
    department_name: str
    capital: str
    latitude: float
    longitude: float
    region_natural: str
    temperature: float
    weather_description: str
    weather_icon: str
    relative_humidity: int
    precipitation: float
    uv_index: float
    wind_speed: float
